canonicalize.go

```diff
@@ -84,8 +84,76 @@ func _canonicalHack(el *etree.Element, seenSoFar map[string]bool) *etree.Element
 // any other other shortcomings until I can get the fixes upstream.
 // NOTE(phoebe): Looks like etree also doesn't remove attributes that are
 // duplicate namespaces. They should be removed if a parent has a matching one
+// NOTE(astuart): namespaces must also be moved to the first element at which they're used
 func canonicalHack(el *etree.Element) *etree.Element {
-	attrMap := make(map[string]bool)
+	attrMap := make(map[string]struct{})
 	return _canonicalHack(el, attrMap)
+}
+
+func getNsDecl(a etree.Attr) string {
+	return fmt.Sprintf("xmlns:%s", a.Key)
+}
+
+type c14nSpace struct {
+	a    etree.Attr
+	used bool
+}
+
+const nsSpace = "xmlns"
+
+func _excCanonicalPrep(el *etree.Element, _alreadyDeclared map[string]c14nSpace) *etree.Element {
+	//Copy alreadyDeclared map
+	alreadyDeclared := make(map[string]c14nSpace, len(_alreadyDeclared))
+	for k := range _alreadyDeclared {
+		alreadyDeclared[k] = _alreadyDeclared[k]
+	}
+
+	usedHere := make(map[string]struct{})
+
+	if el.Space != "" {
+		usedHere[el.Space] = struct{}{}
+	}
+
+	toRemove := make([]string, 0, 0)
+
+	for _, a := range el.Attr {
+		switch a.Space {
+		case nsSpace:
+			toRemove = append(toRemove, a.Space+":"+a.Key)
+			if _, ok := alreadyDeclared[a.Key]; !ok {
+				alreadyDeclared[a.Key] = c14nSpace{a: a, used: false}
+			}
+		default:
+			if a.Space != "" {
+				usedHere[a.Space] = struct{}{}
+			}
+		}
+	}
+	for _, attrK := range toRemove {
+		el.RemoveAttr(attrK)
+	}
+
+	for k := range usedHere {
+		spc := alreadyDeclared[k]
+		//If previously unused, mark as used
+		if !spc.used {
+			el.Attr = append(el.Attr, spc.a)
```

Comment:
Shoot, I should have left in more of my comments. I had to do a lot of reading into [the xml-exc-c14n spec](https://www.w3.org/TR/2002/REC-xml-exc-c14n-20020718/) to figure out how it differs from xml-c14n#, so I had a lot of comments that would probably have been helpful.

Long story short, this step is *only* interested in namespaces. The `alreadyDeclared` and `usedHere` maps track only namespaces so that we can properly exclusively canonicalize those attributes, not declaring them before elements where they're actually used. 

I'll rename the maps to make that clearer.

- Yes, that's what I discovered while reading the spec. exc-xml-c14n removes all namespaces until they're declared, then omits them from all children. It results in duplication sometimes (like in the signature, for one), but it's how the spec states it.
- The indiscriminate removal is for simplicity. Rather than tracking "xmlns we don't need to re-add," I boiled it down to "remove everything, then add *anything* that's used in this context that hasn't been seen before".
- Since we're only concerned with namespaces, I ignore the empty namespace, again for simplicity in the later iteration over `usedHere`.
- If the namespace was previously unused, set `.used` to true, and place it in the map. The map insertion is a bit misleading here, probably. Again, it's for simplicity sake. You can't set a property on an object dereferenced via index (i.e. the following does not compile: `alreadyDeclared["foo"].used = true`). Rather than using pointers and complex value copying syntax, I opted for a map of values that requires an extra assignment to update any item, which we only really do a few times. This could be made to work with pointers, but would still result in many copies being made since each element cares only about its ancestors and not its siblings.

Hopefully that helps!